src/PostgREST/Request/ApiRequest.hs

```diff
@@ -480,37 +484,41 @@ rawContentTypes AppConfig{..} =
+findProc qi argumentsKeys paramsAsSingleObject allProcs contentType isInvPost =
+  case matchProc of
+    []     -> Left $ NoRpc (qiSchema qi) (qiName qi) (S.toList argumentsKeys) paramsAsSingleObject contentType isInvPost
     [proc] -> Right proc
     procs  -> Left $ AmbiguousRpc (toList procs)
   where
-    bestMatch =
-      case M.lookup qi allProcs of
-        Nothing     -> []
-        Just [proc] -> [proc | matches proc]
-        Just procs  -> filter matches procs
-    -- Find the exact arguments match
-    matches proc
-      | paramsAsSingleObject = case pdArgs proc of
-                               [arg] -> pgaType arg `elem` ["json", "jsonb"]
-                               _     -> False
-      | otherwise            = case pdArgs proc of
-                               []   -> null payloadKeys
-                               args -> matchesArg args
-    matchesArg args =
-      -- The function's required arguments are separated from the ones with a default value assigned.
-      -- The set of names of those arguments is compared to the set of keys supplied by the client
-      -- 1. If only required arguments are found, the keys must be exactly the same as those arguments
-      -- 2. If only optional arguments are found, the keys must be a subset of those arguments
-      -- 3. If both required and optional arguments are found, the result of taking away the optional arguments
-      --    from the keys must be exactly the same as the required arguments
-      case L.partition pgaReq args of
-        (reqArgs, [])      -> payloadKeys == S.fromList (pgaName <$> reqArgs)
-        ([], defArgs)      -> payloadKeys `S.isSubsetOf` S.fromList (pgaName <$> defArgs)
-        (reqArgs, defArgs) -> payloadKeys `S.difference` S.fromList (pgaName <$> defArgs) == S.fromList (pgaName <$> reqArgs)
+    matchProc = filter matchesParams $ M.lookupDefault mempty qi allProcs -- first find the proc by name
+    matchesParams proc =
+      let params = pdParams proc in
+      -- exceptional case for Prefer: params=single-object
+      if paramsAsSingleObject
+        then length params == 1 && (ppType <$> headMay params) `elem` [Just "json", Just "jsonb"]
+      -- If the function has no parameters, the arguments keys must be empty as well
+      else if null params
+        then null argumentsKeys
+      -- If the function is called with post and has a single unnamed parameter
+      -- it can be called depending on content type and the parameter type
+      else if isInvPost && length params == 1 && (ppName <$> headMay params) == Just mempty
+        then case headMay params of
+          Just prm | contentType == CTApplicationJSON -> ppType prm `elem` ["json", "jsonb"]
+                   | contentType == CTTextPlain       -> ppType prm == "text"
+                   | contentType == CTOctetStream     -> ppType prm == "bytea"
+                   | otherwise                        -> False
+          Nothing  -> False
```

Comment:
Match the function with the single unnamed parameter and content types.

It's necessary to only consider POST(`isInvPost`) because otherwise we'd also break GET requests that also have overloaded functions with a single unnamed json param. `contentType` isn't enough to determine we're in a POST because a GET can also send a `Content-Type` header despite having no body. Also, we default to `application/json` for all requests included GET.